xdem/dem.py

```diff
@@ -71,6 +72,20 @@ def __init__(self, filename, vref_name=None, vref_grid=None, silent=False, **kwa
         # trying to get vref from product name (priority to user input)
         self.__parse_vref_from_fn(silent=silent)
 
+    def __copy__(self):
```

Comment:
This will yield unexpected outcomes if the `.data` attribute has been changed after loading. It will also not work if the DEM is created in memory (e.g. by `DEM.from_array()`). See my comment in a related issue:

> Right now, I think this works:
> 
> ```python
> import geoutils as gu
> import xdem
> fn_img = xdem.examples.FILEPATHS["longyearbyen_ref_dem"]
> r = gu.georaster.Raster(fn_img)
> dem = xdem.dem.DEM(r.filename)
> ```
> 
> Which yields:
> 
> ```
> No metadata could be read from filename.
> type(dem)
> Out[7]: xdem.dem.DEM
> ```
> 
> It goes through the "init" again, but this is not an issue as those run instantly (manipulating strings). If we had to migrate some more problematic, larger objects, we'd have to use a specific function.
> However we would have a problem if `self.filename` did not exist (a Raster copied in memory for instance). Then everything would fail. I guess it would be useful to store the original filename in `georaster.Raster`, maybe under `._filename`? Thoughts on this? @adehecq is getting better at understanding those underdash attributes :D

That only works assuming the Raster object has not been changed from disk:
```python
In [1]: import xdem

In [2]: import geoutils as gu

In [3]: raster = gu.georaster.Raster("examples/Longyearbyen/data/DEM_2009_ref.tif")

In [5]: raster_downsampled = raster.reproject(raster, dst_res=10)

In [6]: dem_downsampled = xdem.dem.DEM(raster_downsampled.filename)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-b3cc9f821625> in <module>
----> 1 dem_downsampled = xdem.dem.DEM(raster_downsampled.filename)

~/Projects/ETH/GlacioHack/xdem/xdem/dem.py in __init__(self, filename, vref_name, vref_grid, silent, **kwargs)
     59         """
     60 
---> 61         super().__init__(filename, **kwargs)
     62 
     63         if self.nbands > 1:

~/.local/share/conda/xdem/lib/python3.8/site-packages/geoutils/satimg.py in __init__(self, filename, attrs, load_data, bands, as_memfile, read_from_fn, datetime, tile_name, satellite, sensor, product, version, read_from_meta, fn_meta, silent)
    256         """
    257 
--> 258         super().__init__(filename, attrs=attrs, load_data=load_data, bands=bands, as_memfile=as_memfile)
    259 
    260         #TODO: maybe the Raster class should have an "original filename" attribute that doesn't get erased during

~/.local/share/conda/xdem/lib/python3.8/site-packages/geoutils/georaster.py in __init__(self, filename, bands, load_data, downsampl, masked, attrs, as_memfile)
     93         # Don't recognise the input, so stop here.
     94         else:
---> 95             raise ValueError('filename argument not recognised.')
     96 
     97         self._read_attrs(attrs)

ValueError: filename argument not recognised.

In [7]: raster_downsampled.filename
```
To make matters worse, it may sometimes work, but will give unexpected outcomes:
```python
In [16]: raster = gu.georaster.Raster("examples/Longyearbyen/data/DEM_2009_ref.tif")

In [17]: raster.data += 5

In [18]: dem = xdem.dem.DEM(raster.filename)
No metadata could be read from filename.

In [19]: assert np.mean(np.abs(raster.data - dem.data)) < 0.001
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-19-cd99e5be1816> in <module>
----> 1 assert np.mean(np.abs(raster.data - dem.data)) < 0.001

AssertionError:
```
The `.data` attribute of Raster can be modified while retaining the reference to the filename on disk. The `Raster.matches_disk` attribute is not even updated to be `False` (I'll submit an issue for this)!

I think the only reliable approach right now is my first hacky suggestion.

This problem will propagate to the `copy()` method in #46

_Originally posted by @erikmannerfelt in https://github.com/GlacioHack/xdem/issues/43#issuecomment-797300145_

Reply:
I'm not sure I understand the issue for `copy()` (it's a valid one for passing a Raster to a DEM though).
As far as I know, the issue does not propagate to copy because:
- `Raster.copy()` is based on `Raster.from_array()` which create a new memory file based on the array.
- `SatImg/DEM.__copy__()` is based on `super().copy()` from Raster or SatImg, which thus creates that new memory file.
- `SatImg/DEM.copy()` uses the copy package to call the shallow `__copy__()` method (it's just a wrapper).
Can you elaborate on the underlying issue?

Reply:
This is the exact same issue as I commented on in the test file:

But you haven't checked what happens if `img.data` is modified, right? If `img2` is loaded from `img.filename`, then it probably won't contain the in-memory modifications to `img.data`

_Originally posted by @erikmannerfelt in https://github.com/GlacioHack/xdem/pull/46#discussion_r593013514_

Reply:
I will double-check, but this shouldn't be an issue because it's the copy that is passed to the class `__init__()`, and filename is just copied (in `__copy__()`):
```python
new_satimg = super().copy()
new_satimg.filename = self.filename
```

Reply:
I see what you mean! Basically we need to add an assertion in `SatelliteImage.__init__()` and `DEM.__init__()` in case data is already loaded so that it doesn't call from `super()` again.

Reply:
Yeah, that should probably work!

Reply:
But I would suggest adding the test that I suggested (`dem1.data += 5` etc.) to make sure that things work expectedly.

Reply:
But wait... this line in `georaster.Raster.__init__()` should prevent any reloading from filename, no?
```python
 # Or, image is already a Memory File.
        elif isinstance(filename, rio.io.MemoryFile):
            self.filename = None
            self.ds = filename.open()
```
So I'm not sure there's any issue.

Reply:
I'm honestly a little lost now haha. As long as my proposed test passes, I would accept any approach to fix it!